tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -180,9 +182,57 @@ private final class Analyzer(config: CommonPhaseConfig,
     }
   }
 
-  private class ClassInfo(data: Infos.ClassInfo) extends Analysis.ClassInfo {
-    private[this] var _linking = false
-    private[this] var _linked = false
+  private def tryLookupClass(encodedName: String)(
+      implicit from: From): Option[ClassInfo] = {
+    /* There is a significant amount of duplication with lookupClass, but it
+     * is unclear how to factor that out without hurting the performance of
+     * lookupClass, which is the common case.
+     */
+    _classInfos.get(encodedName) match {
+      case Some(info) =>
+        Some(info)
+      case None =>
+        inputProvider.loadInfo(encodedName) match {
+          case Some(data) =>
+            val info = new ClassInfo(data, nonExistent = false)
+            _classInfos += encodedName -> info
+            info.linkClasses()
+            Some(info)
+          case None =>
+            None
+        }
+    }
+  }
+
+  private def lookupClass(encodedName: String)(
+      implicit from: From): ClassInfo = {
+    // See the comment in tryLookupClass about the code duplication
+    _classInfos.get(encodedName) match {
+      case Some(info) =>
+        if (info.nonExistent)
+          _errors += MissingClass(info, from)
+        info
+      case None =>
+        val info = inputProvider.loadInfo(encodedName) match {
+          case Some(data) =>
+            new ClassInfo(data, nonExistent = false)
+          case None =>
+            val data = createMissingClassInfo(encodedName)
+            val info = new ClassInfo(data, nonExistent = true)
+            _errors += MissingClass(info, from)
+            info
+        }
+        _classInfos += encodedName -> info
+        info.linkClasses()
+        info
+    }
+  }
+
+  private class ClassInfo(data: Infos.ClassInfo, val nonExistent: Boolean)
```

Comment:
Factory: 👍 

I think keeping `linkClasses()` separate is desirable, precisely because it highlights the mutability of this algorithm. It is important to think about that mutability because of cycle detection. Cycle detection relies on the mutable aspects and the precise timing of mutations involving `_classInfos +=` and `linkClasses()`.